grakn/rpc/cluster/failsafe_task.py

```diff
@@ -0,0 +1,123 @@
+# under the License.
+#
+import time
+from abc import ABC, abstractmethod
+
+import grakn_protocol.protobuf.cluster.database_pb2 as database_proto
+from grpc import RpcError, StatusCode
+
+from grakn.common.exception import GraknClientException
+from grakn.rpc.cluster.replica_info import ReplicaInfo
+
+
+class FailsafeTask(ABC):
+
+    PRIMARY_REPLICA_TASK_MAX_RETRIES = 10
+    FETCH_REPLICAS_MAX_RETRIES = 10
+    WAIT_FOR_PRIMARY_REPLICA_SELECTION_SECONDS: float = 2
+
+    def __init__(self, client, database: str):
+        self.client = client
+        self.database = database
+
+    @abstractmethod
+    def run(self, replica: ReplicaInfo.Replica):
+        pass
+
+    def rerun(self, replica: ReplicaInfo.Replica):
+        return self.run(replica)
+
+    def run_primary_replica(self):
+        if self.database not in self.client.replica_info_map() or not self.client.replica_info_map()[self.database].primary_replica():
+            self._seek_primary_replica()
+        replica = self.client.replica_info_map()[self.database].primary_replica()
+        retries = 0
+        while True:
+            try:
+                return self.run(replica) if retries == 0 else self.rerun(replica)
+            except GraknClientException as e:
+                # TODO: propagate exception from the server in a less brittle way
+                if "[RPL01]" in str(e):  # The contacted replica reported that it was not the primary replica
+                    print("Unable to open a session or transaction, retrying in 2s... %s" % str(e))
+                    time.sleep(self.WAIT_FOR_PRIMARY_REPLICA_SELECTION_SECONDS)
+                    replica = self._seek_primary_replica()
+                else:
+                    raise e
+                # TODO: introduce a special type that extends RpcError and Call
+            except RpcError as e:
+                # TODO: this logic should be extracted into GraknClientException
+                # TODO: error message should be checked in a less brittle way
+                if e.code() == StatusCode.UNAVAILABLE or "[INT07]" in str(e) or "Received RST_STREAM" in str(e):
+                    print("Unable to open a session or transaction, retrying in 2s... %s" % str(e))
+                    time.sleep(self.WAIT_FOR_PRIMARY_REPLICA_SELECTION_SECONDS)
+                    replica = self._seek_primary_replica()
+                else:
+                    raise e
+            retries += 1
+            if retries > self.PRIMARY_REPLICA_TASK_MAX_RETRIES:
+                raise self._cluster_not_available_exception()
+
+    def run_any_replica(self):
+        if self.database in self.client.replica_info_map():
+            replica_info = self.client.replica_info_map()[self.database]
+        else:
+            replica_info = self._fetch_database_replicas()
+
+        replicas = [replica_info.preferred_secondary_replica()] + [replica for replica in replica_info.replicas() if not replica.is_preferred_secondary()]
```

Comment:
When performing an operation in an "any replica" session, we now give the _preferred secondary replica_ priority over other replicas. It will be contacted first, and the other replicas will only be contacted in the event of a failure.

Reply:
@alexjpwalker we could maybe simplify this a bit by sorting based on `is_preferred_secondary()` boolean value; something like this:

```suggestion
        replicas = sorted(replica_info.replicas(), key=lambda replica: not replica.is_preferred_secondary()]
```

This would put all replicas with `is_preferred_secondary` first in the list

Reply:
I think I prefer the current approach, for two reasons:

- It emphasises that there is precisely one preferred secondary replica
- It's clear - we don't need to think about what the sort order is

Reply:
@alexjpwalker ok I can agree with the argument. It would be more clear by being more verbose.